usaspending_api/common/query_with_filters.py

```diff
@@ -227,33 +247,50 @@ class _RecipientSearchText(_Filter):
+            for filter_value in filter_values:
+                upper_recipient_string = es_sanitize(filter_value.upper())
+                query = es_sanitize(upper_recipient_string)
 
-            else:
-                recipient_search_query.append(recipient_name_query)
+                recipient_name_query = ES_Q("query_string", query=query, default_operator="AND", fields=fields)
+
+                if len(upper_recipient_string) == 9 and upper_recipient_string[:5].isnumeric():
+                    recipient_duns_query = ES_Q("match", sub_awardee_or_recipient_uniqu=upper_recipient_string)
+                    recipient_search_query.append(ES_Q("dis_max", queries=[recipient_name_query, recipient_duns_query]))
+                if len(upper_recipient_string) == 12:
+                    recipient_uei_query = ES_Q("match", sub_awardee_or_recipient_uei=upper_recipient_string)
+                    recipient_search_query.append(ES_Q("dis_max", queries=[recipient_name_query, recipient_uei_query]))
+                else:
+                    recipient_search_query.append(recipient_name_query)
+
+        else:
+            fields = ["recipient_name"]
+            for filter_value in filter_values:
+                upper_recipient_string = es_sanitize(filter_value.upper())
+                query = es_sanitize(upper_recipient_string) + "*"
+                if "\\" in es_sanitize(upper_recipient_string):
+                    query = es_sanitize(upper_recipient_string) + r"\*"
+
+                recipient_name_query = ES_Q("query_string", query=query, default_operator="AND", fields=fields)
+
+                if len(upper_recipient_string) == 9 and upper_recipient_string[:5].isnumeric():
+                    recipient_duns_query = ES_Q("match", recipient_unique_id=upper_recipient_string)
+                    recipient_search_query.append(ES_Q("dis_max", queries=[recipient_name_query, recipient_duns_query]))
+                if len(upper_recipient_string) == 12:
+                    recipient_uei_query = ES_Q("match", recipient_uei=upper_recipient_string)
+                    recipient_search_query.append(ES_Q("dis_max", queries=[recipient_name_query, recipient_uei_query]))
+                # If the recipient name ends with a period, then add a regex query to find results ending with a
+                #   period and results with a period in the same location but with characters following it.
+                # Example: A query for COMPANY INC. will return both COMPANY INC. and COMPANY INC.XYZ
+                if upper_recipient_string.endswith("."):
+                    recipient_search_query.append(recipient_name_query)
+                    recipient_search_query.append(
+                        ES_Q({"regexp": {"recipient_name.keyword": f"{upper_recipient_string.rstrip('.')}\\..*"}})
+                    )
+                else:
+                    recipient_search_query.append(recipient_name_query)
```

Comment:
There is quite a bit of duplicated logic other than field names, some of which seems to be missing for the new Subawards implementation (recipient name period logic and some of the extra sanitization logic).

I'd recommend setting these field names in variables in the condition statement, then using the same code block to handle all query types

Reply:
Done